### DISCUSSION TITLE
JS Plugin: Wait for CSS animations to finish

### DISCUSSION THREAD

**Discussion by kaspost** [swup/swup#557] [Help] [answered]
Basically, I'm coding an app with swup that primarily consists of transitions done through pure css. However, there are a few cases where some css properties need to be calculated through javascript before they are animated in. To solve this, my approach was to add the swup js plugin and use it like this:

  `
  new SwupJsPlugin([
      {
          from: '(.*)',
          to: '(.*)',
          in: prepareTransitions,
          out: prepareTransitions,
      },
  ])
  `

And the prepareTransitions function basically looks like this:

  `
  const prepareTransitions = next => {
      calcAndSetCssProperties();
      next();
  }
  `
  
The problem here becomes that swup doesn't wait for the animations to finish when switching between pages. This obviously has to do with the fact that next is invoked instantly after calcAndSetCssProperties. In order to solve this, is there a swup method that I can utilize to wait for all animations to finish and then invoke the next function?

**Comment by hirasso** [comment, marked as answer]
If using JS Plugin, you'll need to use `setTimeout(next, yourTransitionDurationInMilliseconds)` to wait for your transitions to finish.

Alternatively, you could hook into swup's lifecycle events directly and alter the styles from there (without using JS Plugin):

```js
const swup = new Swup(options);

/**
 * Sets the style on the main transition element
 */
const adjustStyle = (newStyle = {}) => {
  const el = document.querySelector('.transition-main');
  if (!el) return;
  for( const [property, value] of Object.entries(newStyle)) {
    el.style[property] = value;
  }
}
/**
 * Sets the background to blue for the out-animation
 */
swup.on('transitionStart', () => {
  adjustStyle({
    background: "blue"
  })
})
/**
 * Sets the style for the in-animation
 */
swup.on('contentReplaced', () => {
  // set the background to red initially
  adjustStyle({
    background: "red"
  })
  // Fade the background to white on the next frame
  requestAnimationFrame(() => {
    adjustStyle({
      background: "white"
    })
  })
});
```

In your CSS, you would then only need to define the transition and initial value:
```css
.transition-main {
  transition: background 400ms;
  background: white;
}
```

**Reply by daun** [reply]
This is the way to go 🥬

If you're looking to set the timing programmatically, you can update a CSS variable from JS:

```javascript
// in swup event before transition
document.documentElement.style.setProperty('--duration', '200ms');
```

```css
.transition-fade {
  transition: opacity var(--duration, 0ms);
}
```

**Reply by kaspost** [reply]
Thank you for the clear answer!

Your swup event-based approach, instead of using the swup js plugin, seems like the way to go. However, let's say that the javascript dom manipulations and calculations take 500ms to execute. Will swup wait for that to finish before animating in the content? I assume not right?

**Reply by daun** [reply]
@kaspost No, it won't by itself wait for any calculations. There's currently no built-in way to extend the wait time in JS if you're letting swup calculate the CSS transition duration.

**Reply by hirasso** [reply]
If your DOM manipulations are executed synchronously (the default), everything including swup will wait until they are finished:

```js
swup.on('transitionStart', () => {
  executeExpensiveDomManipulations(); // will block everything until it finishes. Swup will pick up the transition duration just after this event
})
```
If they are asynchronous, you'll need to use JS Plugin and a [promise](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise) and call `next` when your manipulations and animations are finished: 
```js
new SwupJsPlugin([
  {
    from: '(.*)', // matches any route
    to: '(.*)', // matches any route
    out: async (next) => {
      await executeAsynchronousThings();
      next();
    },
    in: (next) => next() // immediately continues
  }
]);
```
It's pretty unlikely that you'll need the second approach, though, since DOM manipulations always happen in the main thread and will finish before continuing the code execution.

If you need further help, you should create a reduced test case by forking our [template on CodeSandbox](https://codesandbox.io/s/swup-test-case-template-6ci5si).

@daun please correct me if I'm wrong :)

**Reply by daun** [reply]
@hirasso Swup's events should be blocking, yes.

**Reply by kaspost** [reply]
The function attached to the 'transitionStart' event does not seem to be blocking. In my case, swup instantly proceeds with animating in/out content without waiting for the function to finish. This is the code I used to test this:
```javascript
swup.on('transitionStart', () => setTimeout(() => console.log('dom manipulations finished'), 5000))
```
So it seems like swup's events are not blocking, or am I doing something wrong here?

Then I suppose I will have to resort to using the SwupJsPlugin solution provided by @hirasso in the second code block. However, as we discussed earlier, I will have to include some form of logic to wait for the CSS transitions to finish. Is there a inherit swup method that I can use to retrieve the duration until all the CSS transitions have finished?

If it exists, perhaps it's a good feature to expose as a helper in swup/lib/helpers as well?

**Reply by kaspost** [reply]
Oh, just realized setTimeout is not code blocking. My bad!

I'll respond with a new answer soon but in the meantime is there such a swup method available?

**Reply by daun** [reply]
@kaspost Yeah, `setTimeout` itself is non-blocking. There's [getAnimationPromises](https://github.com/swup/swup/blob/master/src/modules/getAnimationPromises.js) which should do what you're looking for.

**Reply by daun** [reply]
The way you'd use it is: `await swup.getAnimationPromises()`

**Reply by kaspost** [reply]
Tested again with an actual synchronous wait function which proved that swup indeed does wait for the event attached function execution to finish.
```javascript
const syncWait = ms => {
    const end = Date.now() + ms
    while (Date.now() < end) continue
}

const onTransitionStart = () => {
    syncWait(5000)
    console.log('dom manipulations finished')
}

swup.on('transitionStart', onTransitionStart)
```

**Reply by kaspost** [reply]
Awesome, thanks for the help guys!

**Reply by hirasso** [reply]
You are welcome! And thanks for the `syncWait` function, never knew that kind of stuff was possible :)

**Reply by kaspost** [reply]
Yeah, just only use it for testing. Very performance heavy :)

**Reply by hirasso** [reply]
😄 of course!
